resolve TestManifestFilesAreEqual flakes

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120598/pull-kubernetes-unit/1701530801671245824

```
{Failed;Failed;Failed;  === RUN   TestManifestFilesAreEqual/manifests_are_not_equal
    utils_test.go:832: ManifestFilesAreEqual diff doesn't expected
        manifests are not equal
        	expected diff: 
        - 					"2",
        + 					"1",
        actual diff:   &v1.Pod{
          	TypeMeta:   {Kind: "Pod", APIVersion: "v1"},
          	ObjectMeta: {Name: "etcd", Namespace: "kube-system", Labels: {"component": "etcd", "tier": "control-plane"}},
          	Spec: v1.PodSpec{
          		Volumes:        nil,
          		InitContainers: nil,
          		Containers: []v1.Container{
          			{
          				Name: "",
          				Image: strings.Join({
          					"gcr.io/google_containers/etcd-amd64:3.1.1",
        - 					"2",
        + 					"1",
          				}, ""),
          				Command: nil,
          				Args:    nil,
          				... // 20 identical fields
          			},
          		},
          		EphemeralContainers: nil,
          		RestartPolicy:       "",
          		... // 34 identical fields
          	},
          	Status: {},
          }
    --- FAIL: TestManifestFilesAreEqual/manifests_are_not_equal (0.01s)
;=== RUN   TestManifestFilesAreEqual/manifests_are_not_equal_for_adding_new_defaults
    utils_test.go:832: ManifestFilesAreEqual diff doesn't expected
        manifests are not equal for adding new defaults
        	expected diff: 
        - 		RestartPolicy:                 "Always",
        + 		RestartPolicy:                 "",
        actual diff:   &v1.Pod{
          	TypeMeta:   {Kind: "Pod", APIVersion: "v1"},
          	ObjectMeta: {Name: "etcd", Namespace: "kube-system", Labels: {"component": "etcd", "tier": "control-plane"}},
          	Spec: v1.PodSpec{
          		... // 2 identical fields
          		Containers:                    {{Image: "gcr.io/google_containers/etcd-amd64:3.1.11"}},
          		EphemeralContainers:           nil,
        - 		RestartPolicy:                 "Always",
        + 		RestartPolicy:                 "",
          		TerminationGracePeriodSeconds: nil,
          		ActiveDeadlineSeconds:         nil,
          		... // 32 identical fields
          	},
          	Status: {},
          }
    --- FAIL: TestManifestFilesAreEqual/manifests_are_not_equal_for_adding_new_defaults (0.00s)
;=== RUN   TestManifestFilesAreEqual
--- FAIL: TestManifestFilesAreEqual (0.02s)
;}
```

are we catching the flakynes we are seeing in e2e CI with this?
related to:
https://github.com/kubernetes/kubeadm/issues/2927
?

--- a/cmd/kubeadm/app/util/staticpod/utils.go
+++ b/cmd/kubeadm/app/util/staticpod/utils.go
@@ -29,8 +29,8 @@ import (
 	"strings"
 	"sync"
 
-	"github.com/google/go-cmp/cmp"
 	"github.com/pkg/errors"
+	"github.com/pmezard/go-difflib/difflib"
 
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
@@ -374,7 +374,28 @@ func ManifestFilesAreEqual(path1, path2 string) (bool, string, error) {
 	if bytes.Equal(hash1, hash2) {
 		return true, "", nil
 	}
-	return false, cmp.Diff(pod2, pod1), nil
+
+	manifest1, err := kubeadmutil.MarshalToYaml(pod1, v1.SchemeGroupVersion)
+	if err != nil {
+		return false, "", errors.Wrapf(err, "failed to marshal Pod manifest for %q to YAML", path1)
+	}
+
+	manifest2, err := kubeadmutil.MarshalToYaml(pod2, v1.SchemeGroupVersion)
+	if err != nil {
+		return false, "", errors.Wrapf(err, "failed to marshal Pod manifest for %q to YAML", path2)
+	}
+
+	diff := difflib.UnifiedDiff{
+		A: difflib.SplitLines(string(manifest1)),
+		B: difflib.SplitLines(string(manifest2)),
+	}
+
+	diffStr, err := difflib.GetUnifiedDiffString(diff)
+	if err != nil {
+		return false, "", errors.Wrapf(err, "failed to generate the differences between manifest %q and manifest %q", path1, path2)
+	}
+
+	return false, diffStr, nil
 }
 
 // getProbeAddress returns a valid probe address.

--- a/cmd/kubeadm/app/util/staticpod/utils_test.go
+++ b/cmd/kubeadm/app/util/staticpod/utils_test.go
@@ -766,18 +766,19 @@ func TestManifestFilesAreEqual(t *testing.T) {
 			podYamls:       []string{validPod, validPod2},
 			expectedResult: false,
 			expectErr:      false,
-			expectedDiff: string(`
-- 					"2",
-+ 					"1",`),
+			expectedDiff: `@@ -12 +12 @@
+-  - image: gcr.io/google_containers/etcd-amd64:3.1.11
++  - image: gcr.io/google_containers/etcd-amd64:3.1.12
+`,
 		},
 		{
 			description:    "manifests are not equal for adding new defaults",
 			podYamls:       []string{validPod, invalidWithDefaultFields},
 			expectedResult: false,
 			expectErr:      false,
-			expectedDiff: string(`
-- 		RestartPolicy:                 "Always",
-+ 		RestartPolicy:                 "",`),
+			expectedDiff: `@@ -14,0 +15 @@
++  restartPolicy: Always
+`,
 		},
 		{
 			description:    "first manifest doesn't exist",
@@ -830,7 +831,7 @@ func TestManifestFilesAreEqual(t *testing.T) {
 			}
 			if !strings.Contains(diff, rt.expectedDiff) {
 				t.Errorf(
-					"ManifestFilesAreEqual diff doesn't expected\n%s\n\texpected diff: %s\nactual diff: %s",
+					"ManifestFilesAreEqual diff doesn't expected\n%s\n\texpected diff: %s\n\tactual diff: %s",
 					rt.description,
 					rt.expectedDiff,
 					diff,

